crates/kubelet/src/device_plugin_manager/node_patcher.rs

```diff
@@ -0,0 +1,220 @@
+    node_name: String,
+    devices: Arc<Mutex<DeviceMap>>,
+    // Broadcast sender so clonable
+    update_node_status_sender: broadcast::Sender<()>,
+    client: kube::Client,
+}
+
+impl NodeStatusPatcher {
+    pub fn new(
+        node_name: &str,
+        devices: Arc<Mutex<DeviceMap>>,
+        update_node_status_sender: broadcast::Sender<()>,
+        client: kube::Client,
+    ) -> Self {
+        NodeStatusPatcher {
+            node_name: node_name.to_string(),
+            devices,
+            update_node_status_sender,
+            client,
+        }
+    }
+
+    async fn get_node_status_patch(&self) -> NodeStatus {
+        let devices = self.devices.lock().unwrap();
+        let capacity: BTreeMap<String, Quantity> = devices
+            .iter()
+            .map(|(resource_name, resource_devices)| {
+                (
+                    resource_name.clone(),
+                    Quantity(resource_devices.len().to_string()),
+                )
+            })
+            .collect();
+        let allocatable: BTreeMap<String, Quantity> = devices
+            .iter()
+            .map(|(resource_name, resource_devices)| {
+                let healthy_count: usize = resource_devices
+                    .iter()
+                    .filter(|(_, dev)| dev.health == HEALTHY)
+                    .map(|(_, _)| 1)
+                    .sum();
+                (resource_name.clone(), Quantity(healthy_count.to_string()))
+            })
+            .collect();
+        NodeStatus {
+            capacity: Some(capacity),
+            allocatable: Some(allocatable),
+            ..Default::default()
+        }
+    }
+
+    async fn do_node_status_patch(&self, status: NodeStatus) -> anyhow::Result<()> {
+        let node_client: Api<Node> = Api::all(self.client.clone());
+        let _node = node_client
+            .patch_status(
+                &self.node_name,
+                &PatchParams::default(),
+                &kube::api::Patch::Strategic(status),
+            )
+            .await
+            .map_err(|e| anyhow::anyhow!("Unable to patch node status: {}", e))?;
+        Ok(())
+    }
+
+    pub async fn listen_and_patch(self) -> anyhow::Result<()> {
+        // Forever hold lock on the status update receiver
+        let mut receiver = self.update_node_status_sender.subscribe();
+        println!("entered listen_and_patch");
+        loop {
+            println!("listen_and_patch loop");
+            match receiver.recv().await {
+                Err(_e) => {
+                    error!("Channel closed by senders");
+                    // TODO: bubble up error
+                }
+                Ok(_) => {
+                    // Grab status values
+                    let status_patch = self.get_node_status_patch().await;
+                    // Do patch
+                    self.do_node_status_patch(status_patch).await?;
+                }
+            }
+        }
+        // TODO add channel for termination?
```

Comment:
Yeah, I would recommend this. If you add a `start` method to `DeviceManager`, you could add an accompanying `stop` method that uses a `oneshot` channel to inform this method to exit the loop

Reply:
It seems like a termination case is not needed here or for the DeviceManager in general as there aren't clean up steps. The device plugin framework puts the onus on device plugins to re-register on kubelet restarts.